askama/src/filters/humansize.rs

```diff
@@ -43,82 +41,186 @@ impl FastWritable for FilesizeFormatFilter {
         dest: &mut W,
         values: &dyn Values,
     ) -> crate::Result<()> {
-        if self.0 < 1e3 {
+        if self.0 < 1000 {
             (self.0 as u32).write_into(dest, values)?;
             Ok(dest.write_str(" B")?)
-        } else if let Some((prefix, factor)) = SI_PREFIXES
-            .iter()
-            .copied()
-            .find_map(|(prefix_factor, max)| (self.0 < max).then_some(prefix_factor))
-        {
-            // u32 is big enough to hold the number 999_999
-            let scaled = (self.0 * factor) as u32;
-            (scaled / 100).write_into(dest, values)?;
-            format_frac(&mut MaybeUninit::uninit(), prefix, scaled).write_into(dest, values)
         } else {
-            too_big(self.0, dest)
+            // find appropriate unit to format in
+            let unit = SI_PREFIXES
+                .iter()
+                .take_while(|(_, m)| *m <= self.0)
+                .last()
+                .unwrap_or(&SI_PREFIXES[SI_PREFIXES.len() - 1]);
+            // divide number up into "full" units, and remainder
+            let integer = self.0 / unit.1;
+            let remainder = self.0 % unit.1;
+
+            // Compute fractional part with desired precision
+            let precision = self.1.min(15); // max. allowed precision is 15
+            let mut scale = 10u128.pow(precision as u32);
+            let mut fractional = remainder.checked_mul(scale).unwrap_or_default() / unit.1;
+
+            // integer portion (before decimal point)
+            integer.write_into(dest, values)?;
+            if precision > 0 && fractional > 0 {
```

Comment:
`fractional` is only used if the condition `precision > 0` is true. Maybe move the assignment into the if-body?